File: src/main/java/com/zhongan/devpilot/completions/inline/CompletionPreviewInsertionHint.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by FernFlower decompiler)
//

package com.zhongan.devpilot.completions.inline;

import com.intellij.openapi.Disposable;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.LogicalPosition;
import com.intellij.openapi.editor.event.EditorMouseEvent;
import com.intellij.openapi.editor.event.EditorMouseEventArea;
import com.intellij.openapi.editor.event.EditorMouseMotionListener;
import com.intellij.openapi.util.Disposer;
import com.zhongan.devpilot.completions.inline.render.DevPilotInlay;
import java.awt.Component;
import java.awt.Point;
import java.awt.Rectangle;
import java.awt.event.MouseEvent;
import javax.swing.SwingUtilities;

public class CompletionPreviewInsertionHint implements Disposable, EditorMouseMotionListener {
    private Editor editor;
    private DevPilotInlay inlay;
    private String suffix;

    public CompletionPreviewInsertionHint(Editor editor, DevPilotInlay inlay, String suffix) {
        this.editor = editor;
        this.inlay = inlay;
        this.suffix = suffix;
        editor.addEditorMouseMotionListener(this);
        Disposer.register(inlay, this);
    }

    public void mouseMoved(EditorMouseEvent e) {
        if (!this.inlay.isEmpty() && e.getArea() == EditorMouseEventArea.EDITING_AREA) {
            MouseEvent mouseEvent = e.getMouseEvent();
            Point point = mouseEvent.getPoint();
            if (this.isOverPreview(point)) {
                InlineKeybindingHintUtil.createAndShowHint(this.editor, SwingUtilities.convertPoint((Component)mouseEvent.getSource(), point, this.editor.getComponent().getRootPane().getLayeredPane()));
            }
        }
    }

    public void dispose() {
        this.editor.removeEditorMouseMotionListener(this);
    }

    private boolean isOverPreview(Point p) {
        Rectangle bounds = this.inlay.getBounds();
        return bounds != null ? bounds.contains(p) : this.isLogicallyInsideInlay(p);
    }

    private boolean isLogicallyInsideInlay(Point p) {
        LogicalPosition pos = this.editor.xyToLogicalPosition(p);
        if (pos.line >= this.editor.getDocument().getLineCount()) {
            return false;
        } else {
            int pointOffset = this.editor.logicalPositionToOffset(pos);
            Integer inlayOffset = this.inlay.getOffset();
            if (inlayOffset == null) {
                return false;
            } else {
                return pointOffset >= inlayOffset && pointOffset <= inlayOffset + this.suffix.length();
            }
        }
    }
}

File: src/main/java/com/zhongan/devpilot/completions/common/inline/CompletionPreviewInsertionHint.kt
package com.zhongan.devpilot.completions.common.inline

import com.intellij.openapi.Disposable
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.editor.LogicalPosition
import com.intellij.openapi.editor.event.EditorMouseEvent
import com.intellij.openapi.editor.event.EditorMouseEventArea
import com.intellij.openapi.editor.event.EditorMouseMotionListener
import com.intellij.openapi.util.Disposer
import com.zhongan.devpilot.completions.common.inline.render.DevPilotInlay
import java.awt.Component
import java.awt.Point
import javax.swing.SwingUtilities

class CompletionPreviewInsertionHint(
        private val editor: Editor,
        private val inlay: DevPilotInlay,
        private var suffix: String = ""
) : Disposable,
        EditorMouseMotionListener {
    init {
        editor.addEditorMouseMotionListener(this)
        Disposer.register(inlay, this)
    }

    override fun mouseMoved(e: EditorMouseEvent) {
        if (inlay.isEmpty || e.area !== EditorMouseEventArea.EDITING_AREA) {
            return
        }

        val mouseEvent = e.mouseEvent
        val point = mouseEvent.point

        if (!isOverPreview(point)) {
            return
        }

        InlineKeybindingHintUtil.createAndShowHint(
                editor,
                SwingUtilities.convertPoint(
                        mouseEvent.source as Component,
                        point,
                        editor.component.rootPane.layeredPane
                )
        )
    }

    override fun dispose() {
        editor.removeEditorMouseMotionListener(this)
    }

    private fun isOverPreview(p: Point): Boolean {
        return inlay.getBounds()?.contains(p) ?: isLogicallyInsideInlay(p)
    }

    private fun isLogicallyInsideInlay(p: Point): Boolean {
        val pos: LogicalPosition = editor.xyToLogicalPosition(p)

        if (pos.line >= editor.document.lineCount) {
            return false
        }

        val pointOffset = editor.logicalPositionToOffset(pos)
        val inlayOffset = inlay.offset ?: return false

        return pointOffset >= inlayOffset && pointOffset <= inlayOffset + suffix.length
    }
}
